### ISSUE TITLE
Overlapping Images - Bootstrap User

### ISSUE THREAD

**Issue by mattcanty** [desandro/masonry#1147]
**Test case:** https://codepen.io/matthewcanty/pen/NWbEQOj
(you may not see the issue, so I provide screenshot as well 😄 )

<img width="1680" alt="image" src="https://user-images.githubusercontent.com/2089581/110673043-335b5d00-81c8-11eb-8ee5-3691ed8e53d8.png">

Hi, I've been building a website to showcase photography. I used Bootstrap and followed the basic Masonry advice:

```
etc
<div class="row grid" data-masonry='{"percentPosition": true }'>
etc
```

I realise that it's likely due to the image size. But as my JS is minimal (all I did was add the `data-masonry` attribute) I'm not sure now how to fix. I have put a hack in, and it is working.

```js
setTimeout(function () {
  var msnry = new Masonry('.grid');
  msnry.layout();
}, 100);

setTimeout(function () {
  var msnry = new Masonry('.grid');
  msnry.layout();
}, 300);

setTimeout(function () {
  var msnry = new Masonry('.grid');
  msnry.layout();
}, 1000);

setTimeout(function () {
  var msnry = new Masonry('.grid');
  msnry.layout();
}, 5000);
```

The theory is that most will download all the images fairly quickly, especially if cached. Thereforethe 100/300ms take care of those to give a "rapid" feel. For everyone else I through in a 1000 & 5000 for good measure.

Obviously I don't _like_ this, but I will keep it until I have dealt with other issues!

**Comment by rslhdyt** [comment]
same issue with bootstrap 5

**Comment by hamransp** [comment]
Yes, Masonry overlapping item images in bootstrap 5

**Comment by tang2087** [comment]
I faced the same issue as images loading is not completed when the layout is refreshed

**Comment by melihcoban** [comment]
I am facing the same issue with the Bootstrap 5. Applying the mattcanty's hotfix fixes the issue.

**Comment by mattcanty** [comment]
More of a hack than a fix but I’m glad it alleviates the issue right now.

**Comment by tang2087** [comment]
I have adopted mattcanty’s fix and also added another approach of adding event listener on image load:
[Masonry layout](https://kontext.tech/column/xml_xhtml_xslt_dhtml_css_javascript/784/bootstrap-5-masonry-cards-layout) 
You could also use Promise to call layout once all images are loaded. However user experience might not be great as users can view the significant transition from overlapping to a neat layout.

**Comment by thexzan** [comment]
This is the way i solve it, it wait until all images is loaded and then trigger the Masonry
 it may not have the best UX but it work.

```js
Promise.all(Array.from(document.images).filter(img => !img.complete).map(img => new Promise(resolve => { img.onload = img.onerror = resolve; }))).then(() => {
	var msnry = new Masonry('.grid');
	msnry.layout();
});

**Comment by tang2087** [comment]
> This is the way i solve it, it wait until all images is loaded and then trigger the Masonry
> it may not have the best UX but it work.
> 
> ```js
> Promise.all(Array.from(document.images).filter(img => !img.complete).map(img => new Promise(resolve => { img.onload = img.onerror = resolve; }))).then(() => {
> 	var msnry = new Masonry('.grid');
> 	msnry.layout();
> });
> ```

Yes I used the same approach and shared it in my website too. :)

**Comment by martin-0x029A** [comment]
Try with imagesLoaded. First load imagesLoaded.js and masonry.js, and then place the following script at the bottom.


```
<script type="text/javascript">
  
 var $grid = $('#masonry').masonry({
   itemSelector: '.col',
   percentPosition: true
 });

 // layout Masonry after each image loads
 $grid.imagesLoaded().progress( function() {
   $grid.masonry();
 });

</script>
```

Sample HTML:
```
<div id="masonry" class="row row-cols-2 row-cols-sm-3 row-cols-md-4 row-cols-lg-5 g-3 " >
	<div class="col">
               some image
	</div>
     	<div class="col">
               some image
	</div>
	<div class="col">
               some image
	</div>
	<div class="col">
               some image
	</div>
</div>
```

**Comment by taimoor-shan** [comment]
@mattcanty Resolved.
just "replace" async with "sync" in CDN link.
	
<script src="https://cdn.jsdelivr.net/npm/masonry-layout@4.2.2/dist/masonry.pkgd.min.js" integrity="sha384-GNFwBvfVxBkLMJpYMOABq3c+d3KnQxudP/mGPkzpZSTYykLBNsZEnG2D9G/X/+7D" crossorigin="anonymous" sync></script>

**Comment by dreamersofdreams** [comment]
tried a lot of various solutions offered on masonry site, but only thing that worked for me on Bootstrap 5 was using @mattcanty hack above... hopefully this will eventually be addressed
